### DISCUSSION TITLE
Setting the web3.js provider

### DISCUSSION THREAD

**Discussion by Alim32** [WalletConnect/walletconnect-monorepo#2877] [Developer Support] [answered]
Hey,

I was wondering on how to set the web3.js provider from web3modal. Currently when I connect with Wallet Connect it throws an error that the provider is not set. I can't really seem to find what to set it with however.

**Comment by chris13524** [comment]
@Alim32 did/can you follow the guide [here](https://docs.walletconnect.com/2.0/web/web3modal/react/wagmi/installation)? Web3Modal should not require a web3.js provider.

If this doesn't help, please share a reproducible example that demonstrates the error message that you see! 🙏

**Reply by Alim32** [reply]
I originally just used web3.js in vanilla JS + HTML with a cdn for wallet connect v1. Now that v1 is no longer supported I implemented these steps: https://docs.walletconnect.com/2.0/web/web3modal/html/wagmi/installation

When I connect with injection it works fine when doing transaction through the web3.js way, but for wallet connect v1 I had to set a provider.

Like the code below (simplified). Im just not sure if this is still possible with the web3modal route.

```
const WalletConnectProvider = window.WalletConnectProvider.default;
var provider = new WalletConnectProvider({
    rpc: {
        56: "https://bsc-dataseed.binance.org/"
    }
});
await provider.enable();
web3.setProvider(provider);
```

**Reply by glitch-txs** [reply]
Web3modal is opinionated since uses viem/wagmi. You can use ethereum-provider package with web3.js but this package won't show browser wallets like MetaMask on the modal

**Reply by Alim32** [reply]
Ah so ill need to use wagmi's write contract function for that scenario instead?

**Reply by glitch-txs** [reply, marked as answer]
Correct 👌

**Reply by Alim32** [reply]
I added an exception when using wallet connect to use the wagmi's write function and it worked almost the same as with web3js 👍 

thanks for the info.

**Reply by mostafa-rio** [reply]
@chris13524 So what is the Web3Modal solution for projects that have been using Web3Modal and web3js together?
We managed all of our transactions using web3js and the web3js in our project is using Web3Modal provider and now no way to upgrade to V2

**Reply by glitch-txs** [reply]
@mostafa-rio I built a library for this but it's not official from WalletConnect, since there's no current solution but building your own system of browser wallets or migrating to wagmi atm

**Reply by chris13524** [reply]
@mostafa-rio our [docs](https://docs.walletconnect.com/2.0/web/providers/ethereum) say that ethereum-provider is compatible with web3js. It should work. Here is an example: https://github.com/WalletConnect/web-examples/tree/main/dapps/react-dapp-v2-with-web3js

**Reply by Alim32** [reply]
I wrote a generalized function purely for wallet connect and set a variable to it in the non-module to call web3modal write functions. Then in my web3js code I checked if the connection was WalletConnect or not and then either used the writeFunctionWC or the regular web3 variant.


```
writeFunctionWC = async function (ca, abi, functionName, args, value = null) {
    return await CallFunction_WC(ca, abi, functionName, args, value);
}

async function CallFunction_WC(ca, abi, functionName, args, value) {
    try {
        if (value !== null) {
            const { request } = await WagmiCore.prepareWriteContract({
                address: ca,
                abi: abi,
                functionName: functionName,
                account: account.address,
                args: args,
                value: value
            })
            const { hash } = await WagmiCore.writeContract(request)
            console.log(hash);
        } else {
            const { request } = await WagmiCore.prepareWriteContract({
                address: ca,
                abi: abi,
                functionName: functionName,
                account: account.address,
                args: args
            })
            const { hash } = await WagmiCore.writeContract(request)
            console.log(hash);
        }
    }
    catch (error) {
        ShowError(error.message);
    }
}
```
